hail/src/main/scala/is/hail/types/encoded/EArray.scala

```diff
@@ -98,24 +98,58 @@ final case class EArray(val elementType: EType, override val required: Boolean =
       case t: PCanonicalDict => t.arrayRep
     }
 
-    val len = cb.newLocal[Int]("len", in.readInt())
-    val array = cb.newLocal[Long]("array", arrayType.allocate(region, len))
+    val len = cb.memoize(in.readInt())
+    val array = cb.memoize(arrayType.allocate(region, len))
     arrayType.storeLength(cb, array, len)
 
-    val i = cb.newLocal[Int]("i")
     val readElemF = elementType.buildInplaceDecoder(arrayType.elementType, cb.emb.ecb)
 
-    if (!elementType.required)
+    val pastLastOff = cb.memoize(arrayType.pastLastElementOffset(array, len))
+
+    if (elementType.required) {
+      val elemOff = cb.newLocal[Long]("elemOff", arrayType.firstElementOffset(array, len))
+      cb.forLoop(
+        {}, elemOff < pastLastOff, cb.assign(elemOff, arrayType.nextElementAddress(elemOff)), {
+          readElemF(cb, region, elemOff, in)
+        })
+    } else {
       cb += in.readBytes(region, array + const(arrayType.lengthHeaderBytes), arrayType.nMissingBytes(len))
 
-    cb.forLoop(cb.assign(i, 0), i < len, cb.assign(i, i + 1), {
-      val elemAddr = cb.memoize(arrayType.elementOffset(array, len, i))
-      if (elementType.required)
-        readElemF(cb, region, elemAddr, in)
-      else
-        cb.ifx(arrayType.isElementDefined(array, i),
-          readElemF(cb, region, elemAddr, in))
-    })
+      cb.ifx((len % 64).cne(0), {
+        // ensure that the last missing block has all missing bits set past the last element
+//        arrayType.printDebug(cb, array)
+        val lastMissingBlockAddr = cb.memoize(arrayType.firstElementOffset(array) - 8)
+        val lastMissingBlock = cb.memoize(Region.loadLong(lastMissingBlockAddr))
+        cb += Region.storeLong(lastMissingBlockAddr, (lastMissingBlock | (const(-1L) << len)))
+//        arrayType.printDebug(cb, array)
+      })
+
+      def unsetRightMostBit(x: Value[Long]): Code[Long] =
+        x & (x - 1)
+
+      val presentBits = cb.newLocal[Long]("presentBits", 0L)
+      val mbyteOffset = cb.newLocal[Long]("mbyteOffset", array + arrayType.lengthHeaderBytes)
+      val blockOff = cb.newLocal[Long]("blockOff", arrayType.firstElementOffset(array, len))
+      val inBlockIndexToPresentValue = cb.newLocal[Int]("inBlockIndexToPresentValue", 0)
+      val blockOffIncrement = arrayType.elementByteSize * 64
+
+      cb.forLoop(
+        {},
+          blockOff < pastLastOff,
+          cb.assign(blockOff, blockOff + blockOffIncrement),
+        {
+          cb.assign(presentBits, ~Region.loadLong(mbyteOffset))
```

Comment:
Reverse the bytes, right?

Reply:
No, I was wrong about that. `sun.unsafe` uses the system endianness, so little endian, which means no reverse needed.